### ISSUE TITLE
Make base32 CIDv1 the default for go-ipfs

### ISSUE THREAD

**Issue by kyledrake** [ipfs/go-ipfs#4143]
I understand there's a switch to [CIDv1](https://github.com/ipld/cid#cidv1) soon. I think go-ipfs should use lowercased base32 (rfc4648 - no padding - highest letter) as the default multibase.

The reason this encoding is preferable: it's the one encoding that will work with subdomains ([RFC1035](http://tools.ietf.org/html/rfc1035) + [RFC1123](https://tools.ietf.org/html/rfc1123)). The restrictions are: case-insensitive, a-b0-9 and less than 63 bytes.

For a slight increase in length, you reap enormous benefits:

* The ability to do proper security origins for the HTTP gateway with subdomains (cidv1abcde.dweb.link). This is *very important* if we want to handle reports with Google's safe browsing system (which is designed for origins). With the current design, all content is on the same browser origin, and a single phishing/malware report on any of the IPFS gateways (hosted by us or someone else) will make web browsers block every single thing on the origin with a [giant red warning message](https://ipfs.io/ipfs/QmQCGCeWwMMHJxD6ZEhX6LT5F5QwrUMjuv9dHZpcVyR5GT) until it's cleared up with Google (which from experience can take *several days*!)
* Root paths are in the right place, which dramatically improves compatibility with existing web sites that tend to do a lot of this:
 `<img src="/rootimg.jpg">`
* Allows us to register dweb.link (and ipfs.io, etc.) to the [Public Suffix List](https://publicsuffix.org/), which will prevent the sandboxed content from reading/manipulating cookies on the parent domain (and on other cidv1 subdomains).
* Opens up the ability for go-ipfs to do HTTP Host Header parsing and automatic Let's Encrypt support (if we wanted to), so anyone can set up a public IPFS gateway without additional software. Once Let's Encrypt gets their wildcard cert domains shipped (Dec 2017), this could be a fully automated process. Otherwise something like nginx would be needed (I could write an example nginx.conf that people could use for it).

It should use lowercase base32 characters by default, so that it's consistent with subdomain usage (all the browsers will force lowercase). IIRC the RFC doesn't care if it's lowercased, I think people just default to upper case for legacy reasons.

Obviously an abstraction layer could be written that converts between base32 and something else for use with web gateways and then have a different default, but I think it would less confusing for end users to use one default: the one that will let origins in browsers work.

This approach shouldn't be a problem for webextension plugins, but @lidel feel free to chime in.

**Comment by lidel** [comment]
Good arguments, especially the one about [GoogleSafeBrowsing](https://safebrowsing.google.com/)'s false-positives for public gateways 👍 

CIDv1 format is strongly related to discussion at [ipfs/in-web-browsers: Tackle identifying origins with (or without?) fs: paths](https://github.com/ipfs/in-web-browsers/issues/6). 

I was unable to find definitive, final decision  on which exact encoding will be used apart from @lgierth [initially pondering](https://github.com/ipfs/in-web-browsers/issues/6#issuecomment-281568925) "base16 or base32" and @samholmes [suggesting](https://github.com/ipfs/in-web-browsers/issues/6#issuecomment-320536804) base32 with Crockford's Encoding.

Was the decision made elsewhere? 
If not, this ticket provides good opportunity to do so  🔧

**Comment by diasdavid** [comment]
Thank you for creating this issue, @kyledrake. I agree with your proposal, we can take the opportunity that we are bringing CID to the world for the first time to get base32 as the new default.

If not handled internally correctly (i.e using the string format vs the binary format) it will add significant overhead, but that is just something we can change internally to make sure that we use memory efficiently.

**Comment by lgierth** [comment]
> I was unable to find definitive, final decision on which exact encoding will be used apart from @lgierth initially pondering "base16 or base32" and @samholmes suggesting base32 with Crockford's Encoding.

I'm strongly in favour of making base32 the general default encoding for CIDs everywhere. We need base32 for the ipfs:// URL scheme, and it'd suck if people had to deal with different CID encodings, or even have to use converter tools.

**Comment by lgierth** [comment]
And, I think we haven't had any decision on it -- we just sticked with base58 as that was the original encoding used from the beginning.

**Comment by diasdavid** [comment]
> e just sticked with base58 as that was the original encoding used from the beginning.

Yeah, that was pretty much how the _decision_ got made. Still in time to change though.

**Comment by lgierth** [comment]
> Still in time to change though.

Well I'm all for it :):)

**Comment by samholmes** [comment]
### Notes on Base 32 Encoding

What would it take to get a new base added to the [multibase table](https://github.com/multiformats/multibase/blob/master/multibase.csv)? Specifically, what would it take to add [Crockford's Encoding](http://www.crockford.com/wrmg/base32.html) to the table. As of commenting, it appears [RFC4648](https://tools.ietf.org/html/rfc4648#section-6) and [z-base-32](https://en.wikipedia.org/wiki/Base32#z-base-32) are the only base 32 encodings included in the multibase spec.

An added reason to push for Crockford's Base32 is that it meets the same criteria as [Base58Check](https://en.bitcoin.it/wiki/Base58Check_encoding), the base58 encoding Bitcoin uses for bitcoin addressses:

>// Why base-58 instead of standard base-64 encoding?
// - Don't want 0OIl characters that look the same in some fonts and
//      could be used to create visually identical looking account numbers.
// - A string with non-alphanumeric characters is not as easily accepted as an account number.
// - E-mail usually won't line-break if there's no punctuation to break at.
// - Doubleclicking selects the whole number as one word if it's all alphanumeric.

It seems to be like Crockford's Base32 naturally fits the same goals as Base58Check with the added feature of being case-insensitive.

**Comment by lgierth** [comment]
Let's use whatever base32 variant Javascript and (less important) other programming languages use as their default base32. (I assume it's Crockford's)

**Comment by samholmes** [comment]
### Notes on URLs and URI Schemes

From what I can tell, there is no obvious direction for solving issues surrounding URI Schemes and browser origin policies coupled with them thus far. However, [my rough proposal](https://github.com/ipfs/in-web-browsers/issues/6#issuecomment-320536131) is up for further commenting.

However, my inclination is to specify an alternative format and standard from URI. Then, leave it up to implementations to bridge this new format to a purposed URI scheme. Although a hack at the implementation level, it would open up an opportunity to re-think _what a web address could be_. Maybe a multiresource standard should be defined and added to the multiformats basket?

**Comment by samholmes** [comment]
@lgierth I don't know if there is a default base32 encoding in Javascript. If you would consider Javascript's native toString Number method:

```
var a = []; for (var i = 0; i < 32; i++) a.push((i).toString(32))
console.log(a);
// (32) ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v"]
```

It appears toString uses Base 32 Encoding with Extended Hex Alphabet from RFC4648.

Other than this, the Javascript community modules include many variants of base32 encodings; among them is Crockford's. So, it's safe to say that it's not an obscure encoding at the least.

**Comment by kevina** [comment]
@samholmes 

> What would it take to get a new base added to the multibase table?

I am not sold that Crockford's Base32 is not better than than rfc4648 (that we already use the to in the flatfs datastore) but I don't see any problem with adding an entry to the table and implementing it in go-multibase.  Step one would be to open an issue here: https://github.com/multiformats/multibase/issues.

**Comment by kyledrake** [comment]
Support for crockford base32 (base32check?) seems fairly widespread:
* [lua](https://github.com/aiq/basexx)
* [ruby](https://github.com/levinalex/base32)
* [node](https://github.com/agnoster/base32-js)
* [go](https://github.com/richardlehane/crock32)

Since there seems to be a strong preference for it, I hereby revise the proposal to use crockford base32.

**Comment by kyledrake** [comment]
Worth noting is that there's several different flavors of base32, including (my personal favorite) one that Nintendo games used that was [designed to avoid profanity](https://en.wikipedia.org/wiki/Base32#Video_games).

I'm kindof indifferent as to which version gets used. I chose RFC because it's a standard, it's been around a while, nginx-misc-module supports it, and it probably has the widest support across all programming languages. My only strong preference here is that it's a variation most programming languages already support, so we can minimize devs having to re-invent wheels.

Crockford seems to fit the bill more-or-less as well as RFC, which is my rationale for being OK with using it.

@kevina would you have very strong objections to crockford being used by default by go-ipfs with cidv1?

**Comment by kevina** [comment]
@kyledrake concerning some of the issues (in particular the use of cidv1abcde.dweb.link) have a look at https://github.com/ipfs/go-ipfs/issues/1678#issuecomment-157478515 it the full issue rather long but contains lots of useful context in why we currently use /ipfs/Qm.../hash

**Comment by kevina** [comment]
@kyledrake if we switch to using Base32 I do not have any strong objection to using crockford over the RFC one.  The only reason I would chose the RFC one is because it is a standard and more likely to have an implementation available as part of the language.

What I do have an slightly stronger objection to is switching to Base 32 from Base 58 due to the increase in length.  Let's see how this is progress with various proposed changed:

| What  | Length | Increase | Example |
| --- | ---| --- | --- |
| CidV0 | 46 | | `QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn` |
| CidV1 | 49 | +6.5% | `zdj7WbTaiJT1fgatdet9Ei9iDB5hdCxkbVyhyh8YTUnXMiwYi` |
| Base32 | 59  | +28% | `BAFYBEICZSSCDSBS7FFQZ55ASQDF3SMV6KLCW3GOFSZVWLYARCI47BGF354` |
| | | |
| Blank2b-256 | 52 | +13% | `zDMZof1kvswQMT8txrmnb3JGBuna6qXCTry6hSifrkZEd6VmHbBm` |
| Base32 | 62 | +35% | `BAFYKBZACEBUGFUTJIR6QIE7APO5SHPRY32RUWFI762UYTD5G3U2GK7TPSCNDQ` |

So if we ultimately go with using CidV1 using Blank2b-256 and Base32 as the default the length of the Cid string will increase 35%.  That is a non-trivial amount as apposed to the (I think) original plan of switching to CidV1 using same sha256 hash which provides a minor increase of 3 characters or 6.5%.

However, if everyone else is okay with this length increase I am not going to block a move to using Blank2b-256 and/or Base32.

**Comment by samholmes** [comment]
> However, if everyone else is okay with this length increase I am not going to block a move to using Blank2b-256 and/or Base32.

I'm okay with the increase in length. If I am not mistaken, the trade-off would be making it easier to use the same CID within an `/ipfs/<CID>` address and a URL address. 😃

**Comment by lgierth** [comment]
I'm very comfortable trading increased length for increased portability. Not making base32 or base16 the default means that the browser UX of IPFS **will** suffer. Even if we skip the `<hash>.dweb.link` idea, we need base32 CIDs for `ipfs://<hash>`, and not being able to paste CIDs from go-ipfs into the browser would be a little catastrophe :(

> Support for crockford base32 (base32check?) seems fairly widespread:

I'd be more interested in what stdlib-type libraries use, rather than some individual's library. Random data points: golang's [`encoding/base32`](https://golang.org/pkg/encoding/base32/) uses RFC 4648, and the coreutils `base32` command does too.

Could someone check what other important libraries and tools use, so that we get a small survey?

**Comment by lgierth** [comment]
On a different note, we should default to *lowercase* base32 for readability (and of course accept reading both uppercase and lowercase).

**Comment by kevina** [comment]
Any objects I have to the increase length are mild.  It just that things like increase in length can creep up on you and at some point a few years later we stand having keys 2-3 times the length of the original.  Not saying it will happen, but want to explain where my (mild) objection is coming from.

@lgierth why won't our existing base (base58btc) work if you go with `ipfs://<hash>`?  A pointer to another issue documentation is fine.

**Comment by kevina** [comment]
Also, I agree with using lowecase as the default.

**Comment by Stebalien** [comment]
In general, browsers assume that security origins are insensitive. With `ipfs://hash`, `hash` is the security origin.

**Comment by lgierth** [comment]
As per the [WHATWG URL spec](https://url.spec.whatwg.org), the hash in `ipfs://<hash>` is a *domain*, which needs to be a valid *label* according to [RFC 1035](http://tools.ietf.org/html/rfc1035).

That's why @kyledrake made [hshca](https://github.com/neocities/hshca)

**Comment by kevina** [comment]
@Stebalien @lgierth thanks

**Comment by kevina** [comment]
So if we do make base32 the default so that re can represent them on the domain component of the URL the question I have is how will we reference CidV0 objects since we can't completely eliminate them.

I created an issue to discuss this https://github.com/ipfs/go-cid/issues/34.

**Comment by mib-kd743naq** [comment]
I know I am super-late to the party, but is there a reason you are not considering base36 ( as opposed to base32? ) A fully "framed" blake2b 256 bit CIDv1 is 38 bytes long.

In base32 (due to padding specifics) this is 62 characters
In GMP base36 this is `38 * log(256) / log(36)` ~ 59 characters

While the saving of 3 characters is not huge, it's still something... And all the implementations have a GMP engine due to needing to support base58 anyway. An upside is that the question of which base32 padding scheme is used goes away entirely.

**Comment by kevina** [comment]
@mib-kd743naq base 32 is more standard.  Also computing base 32 (and base 64) is far cheaper than bas36 or base58 because non-power of two bases require expensive integer division.  The tiny amount of space saved is not worth it.

**Comment by samholmes** [comment]
@mib-kd743naq I'd argue another reason for base32 over base64: Confusion over look-alike characters (I, L, and O can be confused with number symbols for one and two). This is why base58 was chosen for Bitcoin (https://en.bitcoin.it/wiki/Base58Check_encoding), and I assume the reason why IPFS adopted base58 as well.

My argument for [Crockford's](http://www.crockford.com/wrmg/base32.html) encoding for base32 stands over older encoding standards like RFC 4648. I think that Crockford's encoding is the closest equivalent to the goals for Base58Check while having the added property of being case-insensitive. It's also nice to have digit equivalents shared between decimals and hexadecimals: 0-9 symbols are the same in all three bases, and a-f is the same in base16 and base32 (when base32 is encode using Crockford's encoding).

**Comment by kevina** [comment]
@samholmes I don't think anyone here wants to use Base64, @mib-kd743naq was arguing for base 36.

**Comment by mib-kd743naq** [comment]
Fair enough. My dislike of base32 is that it is an unwieldy encoding, with a pre and post padding scheme, that has to be done using "chunk math". Consider the case of a 38byte CID:
- the first 35 bytes are processed 1:1
- The result needs to be properly 0 ( i.e. `A` )-padded given the CID itself often has leading 0s
- the remaining 3 bytes need to be padded as well
- ugh.

In comparison: with base58 all one has to do is feed the CID value to the hand-optimized GMP binding and that's it.

In any case - 32 it is.

**Comment by kevina** [comment]
@mib-kd743naq we are proposing base32 without any padding characters.  In that light, I am not sure how base32 is an unwieldy encoding compared to say base 36 or base 58.

**Comment by mib-kd743naq** [comment]
@kevina this is the padding I am referring to: https://github.com/whyrusleeping/base32/blob/c30ac30633ccdabefe87eb12465113f06f1bab75/base32.go#L121-L164

**Comment by kevina** [comment]
@mib-kd743naq base32 (when padding characters are not used) doesn't have to be implemented that way.  It could theoretically be implemented in the same way base 36 or base 58 is.

**Comment by samholmes** [comment]
@kevina you missed my point. Let me clarify. One of the reasons why bitcoin used base 58 instead of something like base 64, or base 62, or any base higher than 58 was for human-readability.

**Comment by Kubuxu** [comment]
Also it was very important aspect for Bitcoin as you are visually comparing address when making a transfer.

**Comment by samholmes** [comment]
@kubuxu Visual comparison is important for IPFS. Address spoofing is not an attack vector that should be had in IPFS.

**Comment by lgierth** [comment]
Hosts in URLs are case-insensitive so base64 and base58 are out (as mentioned somewhere above)

**Comment by lgierth** [comment]
I'm gonna make a call here because we need to move forward.

Let's go with @kyledrake's original proposal:

> I think go-ipfs should use **lowercased base32 (rfc4648 - no padding - highest letter)** as the default multibase

If anyone has important reasons not to go with this proposal, please call a veto.

@whyrusleeping @Kubuxu @Stebalien is this still something we can include in v0.4.11?

**Comment by Stebalien** [comment]
The problem here is the CID spec and how go-ipfs implements the datastore.

1. According to the spec, all CIDv0s must be base58.
2. The datastore stores values by CID instead of by multihash so a CIDv1 can't retrieve a block associated with a CIDv0 even if the multihashes match.

Note: this currently "just works" because we ignore 1 in go and allow non-base58 CIDv0s.

**Comment by Stebalien** [comment]
@lgierth see https://github.com/ipfs/go-ipfs/issues/4143#issuecomment-325583021

**Comment by lgierth** [comment]
@Stebalien ah right! Let's defer this to 0.4.12 then

**Comment by olizilla** [comment]
The success of the ipfs-in-web-browsers effort is becoming increasingly dependent on being able to use the CID as the authority in ipfs://<zHash> style addresses. The browsers normalise the authority section to lower-case before we can intercept them in a web-extension; non-base32 encoded CIDs get mangled before we get to them.

Where should I be pushing to help make this a thing @Stebalien @lgierth @kyledrake ?

**Comment by kevina** [comment]
As I see it, the biggest hold up is the handling of CidV0 is base 32.  Things will be easier if we just allow it, but others don't seam to want to, see https://github.com/ipfs/go-cid/issues/34.  If we go this route the actual code will probably take a week for me to do.

If we want to use this opportunity to force the use of CidV1 as others seam to want things become more complicated.  There are complications that and we have not worked out the details.

**Comment by kevina** [comment]
I'm not sure, but maybe this is something that we can work on in the upcoming developers meeting. See https://github.com/ipfs/developer-meetings/pull/16.

**Comment by lidel** [comment]
AFAIK we could introduce an intermediate step of keeping default CID at v0 until backed issues are 9resolved, but changing CIDv1 encoding to  Base32. 

That way if someone decides to opt-in to CIDv1 (`--cid-version`) they will use future-proof base32. 
This will help us with moving forward with creating tests and introducing support in user-facing libs. 

Y/n?

**Comment by diasdavid** [comment]
> That way if someone decides to opt-in to CIDv1 (--cid-version) they will use future-proof base32.
This will help us with moving forward with creating tests and introducing support in user-facing libs.

👍 👍

**Comment by lidel** [comment]
Cool update: Cloudflare officially supports CIDv1B32 in subdomains:
https://blog.cloudflare.com/continuing-to-improve-our-ipfs-gateway/#subdomain-based-gateway

Example: https://bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfyavhwq.cf-ipfs.com/wiki/

**Comment by lidel** [comment]
CIDv1 been around for a while, and CIDv0 causes us more and more trouble:
- people naively parsing them as Multihash instead of a CID – don't want to point fingers, but it's a known problem
- CIDv0 getting [b0rked by browsers like Firefox](https://github.com/ipfs-shipyard/ipfs-companion/pull/824)

I feel we are approaching the point when its less painful to make the breaking change than constantly debug those bugs and as people to use `--cid-version 1`.


@Stebalien @aschmahmann  do you see any hard blockers for flipping `ipfs add` to use  `--cid-version 1` by default?

**Comment by Stebalien** [comment]
Given the fact that users can just "flip it back", no. We _wanted_ to take the opportunity flip other switches, but that's not going to happen any time soon.

**Comment by aschmahmann** [comment]
I don't see any hard blockers with switch to cidv1.

However, since it changes the internal nodes of the DAG and therefore the hash of the entire UnixFS object we should alert people that this is coming down the pipeline. We've been telling people for a while that assuming `ipfs add <file>` will result in the same output every single time across multiple versions of go-ipfs is not dependable, yet some people have been using it to have "psuedo-verifiable gateways" by checking that `curl ipfsgateway.tld/ipfs/{CID} | ipfs add` produces data with the same multihash.

Putting this in the release notes might be sufficient, but some proactive reaching out and/or starting the release issue very early might help.

**Comment by ribasushi** [comment]
No hard opinion on whether the change should be enacted or not, but:

> since it changes the internal nodes of the DAG and therefore the hash of the entire UnixFS object

is only correct for `--cid-version=1`. If one defaults `--upgrade-cidv0-in-output=true` instead, the entire dag remains identical. Bringing this up as it doesn't seem to have been discussed above.

**Comment by Stebalien** [comment]
> is only correct for --cid-version=1. If one defaults --upgrade-cidv0-in-output=true instead, the entire dag remains identical. Bringing this up as it doesn't seem to have been discussed above.

Yes. But that still has some other problems because we _STILL_ haven't switched over to by-multihash blockstores. Once migrations are less painful, I'd like to pull the trigger on this. But it'll take a bit of time.

**Comment by aschmahmann** [comment]
> If one defaults --upgrade-cidv0-in-output=true instead, the entire dag remains identical. Bringing this up as it doesn't seem to have been discussed above.

I was thinking a bit about that, it's true and it might be enough. However, I'm not sure at the moment where users might run into rough edges if we only had `--upgrade-cidv0-in-output=true`. I'm a bit concerned that if there are rough edges that explaining those to users will be awfully difficult to do.

**Comment by RubenKelevra** [comment]
I think it's worth considering changing the default hash algorithm to [blake2b-256](https://github.com/ipfs/go-ipfs/issues/6893) and the default chunker to buzhash.

There's IMHO no downside for changing the chunker to buzhash, but we might gain more deduplication.

And blake2b is much faster and is based on djb's ChaCha crypto functions. It's much faster without relying on hardware features (which is only present on very powerful modern hardware anyway).

**Comment by lidel** [comment]
>   If one defaults `--upgrade-cidv0-in-output=true` instead, the entire dag remains identical. Bringing this up as it doesn't seem to have been discussed above.

IIRC we discussed this in the past: CIDv0-as-CIDv1  is not a good default for `ipfs add` because it proliferates the roundtrip breakage (download "CIDv1", add it to IPFS with `--cid-version 1`, get a different CIDv1) more than the other way around.

By producing CIDv1 with raw leaves by default the number of CIDv0-as-CIDv1-without-raw-leaves gets frozen and does not grow going forward.

I argue that the least painful/confusing way is to do both:
- make `ipfs add --cid-version 1` the default, so `add` produces CIDv1 with raw leaves
  (if someone already use `--cid-version 1` the CID does not change)
- make `--upgrade-cidv0-in-output=true` the default globally
  (so even the legacy `ipfs object`  produces cidv1)

> I think it's worth considering changing the default hash algorithm to blake2b-256 and the default chunker to buzhash.

This would change the CID of people already using `ipfs add --cid-version 1`, but if makes IPFS more useful on desktops and low-powered devices like rpi, then indeed, worth considering. Are there any benchmarks on actual IPFS data to back up the performance claims?

@Stebalien are you still supportive for making this change along with CIDv1 (https://github.com/ipfs/go-ipfs/issues/6893#issuecomment-588346945) or are we waiting for BLAKE 3 (https://github.com/multiformats/go-multihash/issues/121) ?

**Comment by ribasushi** [comment]
I am strongly against adding non-FIPS hashing as defaults. Being a standard means something is a standard: it ends up being used everywhere. Sure today not everything supports accelerated sha2, but in ~5 years everything will. OTOH you are guaranteed that no general purpose chip will ever ship with blake.

We did that with the filecoin chain+file import ( used blake2 ) and it has been an endless source of inconveniences: blake2 in a browser / in js-land in general is *really* "fun". @rvagg can share more.

**Comment by RubenKelevra** [comment]
> > I think it's worth considering changing the default hash algorithm to blake2b-256 and the default chunker to buzhash.
> 
> This would change the CID of people already using `ipfs add --cid-version 1`, but if makes IPFS more useful on desktops and low-powered devices like rpi, then indeed, worth considering. Are there any benchmarks on actual IPFS data to back up the performance claims?
> @Stebalien are you still supportive for making this change along with CIDv1 ([#6893 (comment)](https://github.com/ipfs/go-ipfs/issues/6893#issuecomment-588346945)) or is this out of scope / waiting for BLAKE3 ([multiformats/go-multihash#121](https://github.com/multiformats/go-multihash/issues/121)) ?

## Lowend to mid level desktops: 

![0_SBY9s294vc9tryfr](https://user-images.githubusercontent.com/614929/112893885-85164980-90db-11eb-8aa1-c5a947f82d1b.png)
Source: https://medium.com/asecuritysite-when-bob-met-alice/blake3-3716708235ac

## ARM

For ARM there's [a pretty new implementation leveraging the NEON instruction set](https://lwn.net/Articles/840751/) for blake2b. So blake2b can be pretty fast on ARM now, too.

## Simple go-ipfs Benchmark

With an older mobile CPU (Intel i3-2330M):

<details>

- fetching the data from an SSD
- after dropping the caches
- with an empty flatfs repo
- using `--nocopy` to use a filestore (to avoid having to write them out again)

Statistics on the file sizes of the files:

```
  1k: 104203
  2k:  30552
  4k:  24023
  8k:  21859
 16k:  12903
 32k:   8454
 64k:   4290
128k:   2641
256k:   1684
512k:    691
  1M:    509
  2M:    282
  4M:    164
  8M:    103
 16M:    116
 32M:    109
 64M:    139
128M:     82
256M:     15
512M:      3
  1G:      3
```

</details>

| CID | chunker | hash | time |
| - | - | - | - |
| 1 | buzhash | blake2b | 2m15,586s |
| 1 | buzhash | sha256 | 3m3,144s |
| 1 | size-256 | blake2b | 2m9,384s |
| 1 | size-256 | sha256 | 2m50,372s |
| 0 | size-256 | sha256 | 2m48,741s |


**tl;dr** buzhash+blake2b is 20% faster than just using --cid=1 right now while being more secure and having better deduplication capabilities. Using buzhash instead of the standard chunker would make the process 8% slower.

## Buzhash

Regarding buzhash: I think it's even more important than blake2b, since it allows IFPS to deduplicate files that have their data not perfectly aligned, like VM images as well as database files.

@ribasushi wrote:
> I am strongly against adding non-FIPS hashing as defaults.

Well, we already did. The noise framework uses Blake2b.

@ribasushi wrote:
> Sure today not everything supports accelerated sha2, but in ~5 years everything will. OTOH you are guaranteed that no general purpose chip will ever ship with blake.

I feel comparing SHA-256 to blake2b isn't really fair. It's more a comparison between SHA3 and blake2b and support for SHA3 seems to be still pretty far away.

I think this assumption works for high-end devices, which don't need this. But mid-range and low-end devices will probably still ship with older processors. I bought for example last year a small media computer which has a Celeron J1900 on board. This processor isn't even capable of accelerating AES, while still very powerful and cheap for the task.



> @Stebalien are you still supportive for making this change along with CIDv1 ([#6893 (comment)](https://github.com/ipfs/go-ipfs/issues/6893#issuecomment-588346945)) or is this out of scope / waiting for BLAKE3 ([multiformats/go-multihash#121](https://github.com/multiformats/go-multihash/issues/121)) ?

BLAKE3 is extremely new and did security downgrades to archive a higher performance. I wouldn't trust it for data integrity. Especially since Blake was released like 9 years ago and has a good security record and is based on ChaCha which has 13 years of good security record.

**Comment by Stebalien** [comment]
1. I have no interest in switching the hash function at this point.
2. I'm happy to consider a buzhash based chunker but I'll defer to @ribasushi as the expert. @ribasushi does this make sense as a step forward?

> make --upgrade-cidv0-in-output=true the default globally
(so even the legacy ipfs object produces cidv1)

I'd like to do this, but I'm concerned about doing this _before_ switching to the multihash backed blockstore as the current blockstore _duplicates_ blocks if we fetch them as CIDv1 but stored them as CIDv0 (to make things like pinning work correctly).

The alternative is to try to prioritize the multihash-backed-blockstore work.

**Comment by ribasushi** [comment]
> does this make sense as a step forward?

@Stebalien my take at present is: no, we should not change just yet. It will provide some superficial benefit short term, but is not a good move medium-to-long term. Elaborating:

All my work from last year indicates that a single-pass chunker is woefully inferior to a hybrid-multipass solution: that is `pre-chunk into larger slabs based on commonly-encountered markers, chunk individual slabs with a CDC as the main pass`.

I worry that going with buzzhash ( or any other single-pass CDC ) is that a new closer-but-not-yet-there default will lock us into a local maxima of "kinda sucks but good enough" from which it will be even harder to move away from in the future ( cue current discussion ).

Regarding specifically the buzzhash vs rabin **as paremeterized today** in go-ipfs: rabin outperforms buzzhash in dedup of typical data ( though as mentioned earlier they both fall short of multipass chunkers ).

Given various current in-flight priorities, the state of UFSv2, upcoming not-yet-explored use-cases of scientific data onboardign via Estuary, it makes sense to punt this change, and just do the b32 switch.

**Comment by ribasushi** [comment]
> The alternative is to try to prioritize the multihash-backed-blockstore work.

As a separate issue: If the above ends up happening anyway, then we are back to `--upgrade-cidv0-in-output=true` being viable. I am quite dubious effectively forcing a recode of all leaves to go from pb-wrapped to raw is worth the churn ( we are *not* explicitly doing this, but the public takeaways/practical outcome of *any* switch of defaults will be "crap, we need to reimport stuff" )

**Comment by aschmahmann** [comment]
>  I am quite dubious effectively forcing a recode of all leaves to go from pb-wrapped to raw is worth the churn

Can you explain this a bit more? Are you saying you'd prefer leaving everything as-is, but using `--upgrade-cidv0-in-output=true` if we were to first land the multihash keyed blockstore work?

I'm not sure how much of an option this is if we want to make this switch soon since it's effectively a network-wide change. i.e. if there is only one user who has the bytes behind a CID that you're looking for and they haven't upgraded their blockstore then `--upgrade-cidv0-in-output=true` will cause you to not get the data.

> but the public takeaways/practical outcome of any switch of defaults will be "crap, we need to reimport stuff"

I think whether people feel the need to reimport something should be dependent on what they're going to get out of the reimport. Examples include:
- Better chunking schemes -> better deduplication
- Better UnixFS format -> better performance, more features, etc.
- Raw-leaves -> less space wasted, easier to map into existing data, potentially future access to features like https://github.com/protocol/beyond-bitswap/pull/29, etc.
- All nodes (particularly the internal nodes) as CIDv1 -> less thinking about encoding changes particularly for web browsers when using a short hash, like the default SHA2-256, and the default base32 encoding

This might be something for us to battle with now at a smaller scale instead of waiting for UnixFSv2 to have that battle.

For example, if all anyone was getting from a reimport was CIDv1 internal nodes then that seems like a bad reason to reimport. If it was just raw-leaves then maybe people would want to reimport if they felt the gains were worthwhile to them (e.g. they want to more heavily use the filestore) - although I'd probably wait for a reimport until something like https://github.com/protocol/beyond-bitswap/pull/29 was implemented to make storing two copies of the data worthwhile.

None of this should change us from flipping the defaults though. CIDv1 internal leaves costs a couple bytes to make life easier, and raw-leaves is just absolutely better. Shouldn't we want our users to import things the better thing by default?

As I mentioned above IMO anyone who is currently assuming `ipfs add <file>` should always produce the same result across every version of go-ipfs is working under an incorrect assumption, defaults should be allowed to change in order for us to make progress. Anyone relying on that behavior will need to be moved off of it and perhaps just changing the default to `--cid-version=1` is enough of a push to move things along here (and people can move back to `--cid-version=0` in the meanwhile if they need to).

Note: `ipfs add --cid-version=1` implies `--raw-leaves` by default so if flipping the cid-version then that's enough to deal with the raw-leaves default https://github.com/ipfs/interface-go-ipfs-core/blob/b935dfe5375eac7ea3c65b14b3f9a0242861d0b3/options/unixfs.go#L106

**Comment by ec1oud** [comment]
It's unbelievable that this is still being discussed with no resolution in sight.  cidv0 is broken: it can't be used in URLs because it's case-sensitive.  Therefore it's in no way suitable for Web 3.0 and was a mistake from the beginning.  So, I'd be glad never to see a Qm cid again.  Instead what I'm seeing is even in 0.8.0, ipfs add is giving me those by default.  Even if the daemon is started with --cid-base=32.  There is no "ipfs daemon" option --cid-version=1 AFAICT.  ~/.ipfs/config has no cid options at all, and the list of bootstrap peers still have Qm CIDs too.  You still expect every individual command-line operation to need a non-default _option_ to get cidv1, in 2021?!?  And if technical debt is holding this back, why isn't it a high priority to get that fixed?  It's been 4 years.

**Comment by RubenKelevra** [comment]
I agree that it's a necessity to change the CID to v1.

I still think changing the chunker to buzzhash makes sense - there's no way we can deduplicate sensible any large files without it, like ISOs or VM files.

Having buzzhash as default isn't optimal, but it's a good baseline we can change in the future - when the future is ready.

The fixed-size-chunks have zero advantage over a rolling hash algorithm.

Blakeb2-256 just makes sense when we target mobile devices and low cost servers, like Raspberry PIs.

Switching to SHA3 is just way too much overhead IMHO and even SHA256 is quite slow in comparison - and offers less security. 

Correct me if I'm wrong, but does switching the CID but not switching the Hash algorithm to blakeb2-256 has any advantage?

**Comment by rvagg** [comment]
> does switching the CID but not switching the Hash algorithm to blakeb2-256 has any advantage?

Browsers is probably the most compelling reason here, Blake2b in JS is a pain in the backside compared to SHA2-256 (many of us are still annoyed at Filecoin choosing that default!), we have native (and efficient) SHA2-256 everywhere JS runs, so it's a nice default.

If someone wants to do some perf comparisons of SHA2-256 (https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest) and the best JS Blake2b hasher they can find then maybe that would help change the needle a little? Or maybe such data exists and it's a matter of digging it up, I'm sure someone with a Blake2b JS library has some numbers if they have actual efficiency gains.

**Comment by BigLep** [comment]
Personal notes for the value of this from verbal conversation:
1. Easy copy/paste into web browsers and avoiding nuances around operating system case sesitivities.
2. Forcing function for the community to support CIDv1 (reducing time spent in future answering questions of ambiguity)
3. We've defaulted to using raw leaves than protobuffs.  More portability, etc.

**Comment by RubenKelevra** [comment]
> > does switching the CID but not switching the Hash algorithm to blakeb2-256 has any advantage?
> 
> Browsers is probably the most compelling reason here, Blake2b in JS is a pain in the backside compared to SHA2-256 (many of us are still annoyed at Filecoin choosing that default!), we have native (and efficient) SHA2-256 everywhere JS runs, so it's a nice default.
> 
> If someone wants to do some perf comparisons of SHA2-256 (https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest) and the best JS Blake2b hasher they can find then maybe that would help change the needle a little? Or maybe such data exists and it's a matter of digging it up, I'm sure someone with a Blake2b JS library has some numbers if they have actual efficiency gains.

I don't have an overview of JS and what's included. But basically, every other major language just includes this in the standard library for crypto. So kind if strange that this is so hard in JS.

But if Filecoin already leads the way in this direction, I think this still might be the way to go - since we have the dependency basically anyway.

Apart from this blake2b is already part of go-ipfs, so data might already arrive at a JS client with this hash. Not being able to read it is quite bad compatibility issue.

**Comment by lidel** [comment]
- go-ipfs [0.12](https://github.com/ipfs/go-ipfs/releases/tag/v0.12.0) shipped the blockstore migration from full CID to Multihash keys, removing the final obstacle. 
- we can now switch produced CIDs to be CIDv1 by default!  https://github.com/ipfs/specs/issues/247#issuecomment-1092314534

@schomatis  mind prioritizing this one and picking it up after you finish the current work? 

This is an important UX change that is a long time due. 
The entire ecosystem will be grateful for cleaning this up and producing CIDv1 by default. :pray: 

Specific tasks would be  to switch below places to create and output CIDv1 (instead of the current CIDv0):

- [ ] MFS  (`ipfs files` + default empty MFS root created on `ipfs init`)
- [ ] legacy `ipfs object new|put|patch` (for completeness, otherwise we will see people producing CIDv0)
- [ ] `ipfs add` – poc in https://github.com/ipfs/go-ipfs/pull/8185 (just a demo, ok to close it and open a new PR)


My thoughts:
- This is a bigger chunk of work,  but  well scoped, and you could work on this without being blocked by others, or waiting for reviews.
- These commands already have tests, after you flip default they need to be refactored to use CIDv1. 
- Given that `ipfs add` breaks most of the tests, you can focus on other tasks first.
- Fine do it all in a single PR or have separate ones for MFS, add, and other commands. Whatever works better for you.

**Comment by olizilla** [comment]
@lidel If i'm reading this right, the greatest change since the invention of CIDs could be scheduled to land in go-ipfs 0.13... an obviously haunted release number. In light of this, the next release of go-ipfs shoud be v1 for parity with the CID version number. 😎

**Comment by schomatis** [comment]
@lidel Can I have a simple example of what would a CID change look like in `ipfs files`, please? Like what is the current behavior and what would you expect after the change. I'm still grappling with what are we aiming for in terms of UX.

**Comment by schomatis** [comment]
(Trying to interpret the spirit of the change, but can be way off. This is just something concrete to start the discussion from.) In the case of `ipfs init` for example, the CID that gets stored in the datastore under `/local/filesroot` is an empty UnixFS directory created with the `go-merkledag` tooling [`NodeWithData`](https://github.com/ipfs/go-merkledag/blob/0545ec1945ce3a33014845cbb74ad42d311ddc7a/node.go#L113-L116
), which creates a `ProtoNode` without a CID builder. This in turn is interpreted by its `CidBuilder()` as using [CIDv0 by default](https://github.com/ipfs/go-merkledag/blob/0545ec1945ce3a33014845cbb74ad42d311ddc7a/node.go#L87-L93) if it hasn't been specified.

We would then want to refactor that library to use CIDv1 by default? Or am I going too deep in the change and instead we would need to find a bypass in MFS/UnixFS to avoid this tooling and make sure they set the CIDv1 everywhere? The first seems the cleanest if we're transitioning to CIDv1 everywhere but might cause unexpected behavior to other consumers. The second has less impact but will be messier and harder to maintain (thinking off the top of my head, need to look closer into the code).

I can still be going too deep and we just need to find a spot to make the CIDv0-to-CIDv1 change in the command output itself, and still remain with CIDv0 internally (this also goes in the direction of an even less impactful change, but again with the sense of a messier approach, at least at first sight, of having to catch and intercept these exposures of our internal CIDs to change them for the user).

**Comment by lidel** [comment]
@schomatis  thank you for looking into this. 

Switching to `v1CidPrefix` in [`CidBuilder()`](https://github.com/ipfs/go-merkledag/blob/0545ec1945ce3a33014845cbb74ad42d311ddc7a/node.go#L87-L93) sounds like a good way to ensure the ecosystem moves to CIDv1, but could also block go-ipfs work for a while.    Let's  discuss this approach independently in https://github.com/ipfs/go-merkledag/issues/86.

A less invasive approach is to call `SetCidBuilder` after `NodeWithData` is created by go-ipfs. We already do that in some places like  [`core/commands/files.go`](https://github.com/ipfs/go-ipfs/blob/master/core/commands/files.go) – we should have a "default builder" hardcoded in go-ipfs somewhere, so we are independent of default in go-merkledag.

**Comment by lidel** [comment]
Relevant libraries moved to boxo and have clear maintainers and ownership:
- https://github.com/ipfs/boxo/tree/main/ipld/unixfs
- https://github.com/ipfs/boxo/tree/main/ipld/merkledag


We should be able to update defaults to CIDv1 with raw-leaves=true as the new default there.

**Comment by lidel** [comment]
Switching to `--cid-version 1` by default implies enabling `--raw-leaves` by default, which will change the DAG chunks and the multihash of the final CID.

It feels like a good opportunity to ALSO enable `--inline` with `--inline-limit=32`, to have one release with "breaking change" instead of two.

**Comment by lidel** [comment]
https://github.com/ipfs/kubo/pull/10421 got merged and is scheduled to ship with Kubo 0.29.

It introduced [`Import` configuration section](https://github.com/ipfs/kubo/blob/master/docs/config.md#import) which allows customizing CID and UnixFS flags for commands like `ipfs add`.

There are also two new profiles with cidv0 and cidv1 presets:

-  `ipfs config profile apply legacy-cid-v0` will make `ipfs add` always produce CIDv0 (solidifying current behavior). 
- `ipfs config profile apply test-cid-v1` allows switching to CIDv1 by default with 1 MiB block size (potentially the future default, matching CIDs produced by services like https://web3.storage, and cutting down the number of CIDs that need to be announced to Amino DHT)


Feedback about what should go into `test-cid-v1` is welcome. Right now, it sets four flags:

https://github.com/ipfs/kubo/blob/099ce9c8f274588739ea11100eb52fb6fbd726a0/config/profile.go#L222-L225

**Comment by bumblefudge** [comment]
> Feedback about what should go into test-cid-v1 is welcome

I'm trying to solve this on two different levels.  On the _spec_ level, I feel like every single parameter of DAG production needs to be defined for the kubo-CIDv1 and kubo CIDv0 profiles, _even if the CLI doesn't expose each of those parameters as flags_ (it's totally OK if it never does!). On the _kubo_ level, it's probably better if many of these parameters _aren't_ exposed as flag, and if special-snowflake IPLD use cases that change some of these low-level DAG-shape settings _cannot_ recreate their CIDs with kubo from the same inputs.  I'm hoping the IPIP version of the profile lists all these "implicit" (in kubo, at least) DAG params, and that the formal, independently-reimplementable definition of these profiles includes their values.

**Comment by lidel** [comment]
Yes, I think what we want to do in Kubo, is to "freeze defaults" by writing them to JSON config (`Import` section) when initial `ipfs init` is called, and also for existing users who have it missing from config.

This way we can switch new instances to CIDv1 without impacting existing users.

Going forward, we if we make any new knob that impacts produced CID, we would add it there and also freeze the default. 

This way users who are oblivious to all the complexity will still get "the same CID" and if they start getting different one on a different Kubo instance, they can DIFF config, spot the difference, and fix  problem themselves.

**Comment by lidel** [comment]
The original ask had two parts:

1. Use CIDv1 with base32 -- available now via `ipfs config profile apply unixfs-v1-2025`. The gateway side (subdomain isolation, PSL registration, origin-based security) has been shipping for years.
2. Make it the default -- new `ipfs init` still defaults to CIDv0 for backward compatibility. Existing users who want CIDv1 can apply the `unixfs-v1-2025` profile or adjust [`Import.*` configuration options](https://github.com/ipfs/kubo/blob/master/docs/config.md#import) manually.

All the technical prerequisites listed throughout the thread (blockstore handling both CID versions, base32 display, subdomain gateways) are done. Various details are documented in ipfs/specs#499

The remaining question is a policy decision about when to flip the default for new nodes from implicit [`unixfs-v0-2015`](https://github.com/ipfs/kubo/blob/master/docs/config.md#unixfs-v0-2015-profile) to implicit [`unixfs-v1-2025`](https://github.com/ipfs/kubo/blob/master/docs/config.md#unixfs-v1-2025-profile). For now we keep v0 as default. We will switch to v1 in "Kubo v1.0" release in the future, to be more prominent and once v1 is dogfooded bit more.

Keeping this open to track when that happens.

**Comment by Chara-Freedom** [comment]
> The remaining question is a policy decision about when to flip the default for new nodes from implicit [`unixfs-v0-2015`](https://github.com/ipfs/kubo/blob/master/docs/config.md#unixfs-v0-2015-profile) to implicit [`unixfs-v1-2025`](https://github.com/ipfs/kubo/blob/master/docs/config.md#unixfs-v1-2025-profile).

Well you asked opinion from the people and I said that the right time is 0.40

<img width="1248" height="356" alt="Image" src="https://github.com/user-attachments/assets/59898980-2096-4721-b1e2-1234ae63b93b" />

Implemented feature should become the default immediately otherwise there will be more discrepancy between people using v0 and v1. I used v1 from the start but even if I did use v0 it wouldn't be painful to switch. As I see it's been 10 years what is more right time to switch? I think it should be reversed - if people want legacy behavior they should clearly choose it, not otherwise. So much work was done by you and you say it's not a right time, while this issue is 8 and a half years old. Well its you and your team who decide this.

**Comment by Chara-Freedom** [comment]
> Keeping this open to track when that happens.

Probably making some survey is also an option.

**Comment by Chara-Freedom** [comment]
Oh I get it now it is a big milestone that can't be done in 0.40 it will be 1.0 😅

**Comment by davidebeatrici** [comment]
I would like to report that, even with the `unixfs-v1-2025` profile applied, files written into the MFS through FUSE get a CIDv0.

Is it a known issue?

**Comment by lidel** [comment]
@davidebeatrici i dont believe so, please fill a new issue so we can track that edge case. provide as much detail as you can (what is the directory structure, are you writing  to top level MFS or a subdirectory, are you using `ipfs files write` or `ipfs add` with `--to-files` etc. Thanks in advance 🙏 

ps. there is manual  `ipfs files chcid --help` you can use as a workaround, but please fill issue as lack of automatic change of this this seems to be an UX gap.

**Comment by davidebeatrici** [comment]
Thanks for the quick reply!

I created #11247.
